### DISCUSSION TITLE
Negative dPhi values

### DISCUSSION THREAD

**Discussion by TheNablaOperator** [gimli-org/pyGIMLi#375] [Q&A]
Hello everyone.
I am still new to pg and tryed to re-calculate old simulation data with gimli.
Initially I made a synthetic model with Geotomo Res3DMod, currently I created in gimli again. 
I calculated it in 2D, 3D as a world model and 3D as a tank model, just to play around and get started with gimli.

With the 3D tank model, 6 survey lines on top in wenner configuration, I have the following issue: 
I end up with negative dPhi values... 

Is that a problem? Concerning about the result...
And what could/should I do to avoid that? Changing the set of parameters?

Thank you in advance.

```
10/03/22 - 07:58:43 - pyGIMLi - INFO - Found 2 regions.
10/03/22 - 07:58:43 - pyGIMLi - INFO - Region with smallest marker (1) set to background
10/03/22 - 07:58:45 - pyGIMLi - INFO - Creating forward mesh from region infos.
10/03/22 - 07:59:08 - pyGIMLi - INFO - Creating refined mesh (H2) to solve forward task.
10/03/22 - 08:02:31 - pyGIMLi - INFO - Use median(data values)=124.67310926125737
10/03/22 - 08:02:31 - pyGIMLi - INFO - Created startmodel from forward operator: 2536 [124.67310926125737,...,124.67310926125737]
10/03/22 - 08:02:31 - pyGIMLi - INFO - Starting inversion.
fop: <pygimli.physics.ert.ertModelling.ERTModelling object at 0x000002074AAFF4A0>
Data transformation: <pgcore._pygimli_.RTransLogLU object at 0x00000207093DFDB0>
Model transformation (cumulative):
	 0 <pgcore._pygimli_.RTransLogLU object at 0x000002074AAF09A0>
min/max (data): 100/184
min/max (error): 3%/3%
min/max (start model): 125/125
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
inv.iter 2 ... chi² = 7.43 (dPhi = 16.94%) lam: 20.0
--------------------------------------------------------------------------------
inv.iter 3 ... chi² = 7.61 (dPhi = -2.36%) lam: 20.0
--------------------------------------------------------------------------------
inv.iter 4 ... chi² = 8.23 (dPhi = -8.11%) lam: 20.0
################################################################################
#                Abort criteria reached: dPhi = -8.11 (< 2.0%)                 #
################################################################################
10/03/22 - 10:55:52 - pyGIMLi - INFO - Saving resistivity data to: .\20220310-10.55\ERTManager
```

**Comment by halbmy** [comment]
What is the chi-square value of the starting model and the first iteration? (Why don't we see it in the log?) Did you use `invert(verbose=True)`?

In my experience, an increasing objective function is an indication of forward modelling errors coming in, either as a result of inconsistent data or due to a too low regularization. Without seeing the script it is hard to speculate.

**Reply by TheNablaOperator** [reply]
Hello. Thanks for the reply.
I don't know... It's the total Log after 

```
mgr = ert.ERTManager(survey)
inv = mgr.invert(verbose=True, 
                 mesh=mesh, 
                 lam=20, 
                 zWeight=0.3) 
```


Yes, `invert(verbose = True, ...)`

Here is the code of the tank model. It is quite the same as the tank example. I changed the sensors and geometry.
Scheme File: [scheme3.shm.txt](https://github.com/gimli-org/gimli/files/8242905/scheme3.shm.txt)

```Python
import matplotlib.pyplot as plt
import pygimli.meshtools as mt
import numpy as np
import pygimli as pg
from pygimli.physics import ert

print(pg.__version__)

quad = mt.createCube(start=[-1, -1, 0],
                     end=[70, 16, -12],
                     boundaryMarker=1)

block = mt.createCube(start=[36, 0, -3],
                     end=[66, 15, -9],
                     marker=2)#,
                     #boundaryMarker=10)

tank = block + quad 

#Resistivity Map
resmap = [[1, 100], [2, 1000]]

tank.exportVTK("tank")
pg.show(tank)

# Open Sensor File with 6 survey lines, 24 elec each line, elec. dist = 3m, pseudo-3D, ERT
shm = pg.DataContainerERT("scheme3.shm")

# Todo:
# Alternatevly create sensor scheme manually.
# Insert example code here, change Sensor Positions with setSensors

for s in shm.sensors():
   # tank.createNode(s, marker=-99)
    tank.createNode(s - [0, 0, 1e-3/2]) #1e-3/2

# Due to N.BC.
tank.createNode([0, 7.5, -6], marker=-999)
tank.createNode([69, 7.5, -6], marker=-1000)

# Refinement for numerical accuracy
for s in tank.positions(pg.find(tank.nodeMarkers() == -99)):
    tank.createNode(s - [0.0, 0.0, 1e-3/2]) #1e-3/2

# Also refine the reference node
tank.createNode([69, 15, -12 - 1e-3/2]) #1e-3/2

# create mesh
mesh = mt.createMesh(tank)
#mesh.exportVTK("tank_mesh.vtk")
print(mesh)
pg.show(mesh, notebook=True)

# Forward Sim
survey = ert.simulate(mesh, 
                   res=resmap, 
                   scheme=shm, 
                   verbose=True) #noiseAbs=1e-6, seed=1337
#, sr=False, calcOnly=True)

print(survey)
pg.show(survey, cMin=100, cMax=300)
survey.save("Forward_Tank_Sim.ohm", 'a b m n rhoa') # a b err k m n u rhoa

# Error estimates
survey["err"] = ert.estimateError(survey)

# Inversion
mgr = ert.ERTManager(survey)
inv = mgr.invert(verbose=True, 
                 mesh=mesh, 
                 lam=20, 
                 zWeight=0.3) #, MarquardtScheme = 0.9)

mgr.saveResult()
pg.show(mgr.paraDomain, mgr.model, label=pg.unit('res'))
```

**Comment by halbmy** [comment]
Why did you comment out the electrode nodes?
```
for s in shm.sensors():
   # tank.createNode(s, marker=-99)
    tank.createNode(s - [0, 0, 1e-3/2]) #1e-3/2
```    
I guess they are responsible for the lacking accuracy. Furthermore you should demand a certain quality, e.g. `createMesh(plc, quality=1.3)` and possibly consider using quadratic shape functions (`mesh.refineP()`).

**Comment by carsten-forty2** [comment]
Please note, you should never run the Simulation and Inversion on the same mesh, its an "inverse crime". Better create a special mesh for inversion without the structural information of your predefined model. Your simulation mesh consists of two markers for `quad` and `block` which will be reinterpreted by the inversion manager as background region  (marker==1) and parametric domain(marker==2 to be inverted and shown as the `paraDomain` mesh). Im not sure if you intented to achieve a homogeneous inversion model for the `block` only.  

I think the electrode marker nodes are not needed because the refinement nodes are close enough below the surface. If I remember right the last similar issue, these electrode nodes are directly aligned on the interface line between `quad` and `block`, which will fail the meshing process. I did not tried your example to lock at the mesh, but if the electrodes are still along the border between these two bodies, you will create a strong 3 dimensional problem which can't be resolved by 2 dimensional measurement. This may lead to bad convergence in the inversion. 

Also you should always add some noise to the synthetic data, because its needed by the inversion theory. If the noise is in the order of the accuracy of the forward simulation, you should achieve a convergence of the inversion process to chi²=1.

**Comment by TheNablaOperator** [comment]
First of all, thank you both for the answers!

@carsten-forty2 I didn't think about the mesh. But yes, it makes more sense to create a new mesh on the simulated data. No, the inversion should run through the total "tank".
Yes in the issue #369 we had the discussion about the mesh mismatch. Setting the sensors right on top of the boundary, ended in a total distorted mesh for the forward model, I already corrected that here... The forward model and its mesh looks good here. 
I already added the noise in the code, I commented it out, when I started debugging...

@halbmy I commented that out, since I had the above mentioned issue. 

I guess creating a new mesh solves the issue, but I am not sure at all. 
However, I have to adapt the meshing routine, since this is a tank problem not a world problem, then I can say something about the inversion result.

**Comment by TheNablaOperator** [comment]
Hello I made a new mesh for the inversion, and get an inversion result.
After 2 iterations I get: chi² = 0.47, dPhi = 38.88%
I am not quite sure if this OK now... ?
But I think it will not get better, since I only have 6 survey lines for the ERT on top of the tank...?!

```
16/03/22 - 09:20:03 - pyGIMLi - INFO - Found 1 regions.
16/03/22 - 09:20:03 - pyGIMLi - INFO - Creating forward mesh from region infos.
16/03/22 - 09:20:13 - pyGIMLi - INFO - Creating refined mesh (H2) to solve forward task.
16/03/22 - 09:21:04 - pyGIMLi - INFO - Use median(data values)=125.13665600253084
16/03/22 - 09:21:04 - pyGIMLi - INFO - Created startmodel from forward operator: 95979 [125.13665600253084,...,125.13665600253084]
16/03/22 - 09:21:04 - pyGIMLi - INFO - Starting inversion.
fop: <pygimli.physics.ert.ertModelling.ERTModelling object at 0x0000020E6D3DD590>
Data transformation: <pgcore._pygimli_.RTransLogLU object at 0x0000020E6D33E630>
Model transformation (cumulative):
	 0 <pgcore._pygimli_.RTransLogLU object at 0x0000020E6D345100>
min/max (data): 100/184
min/max (error): 3%/3%
min/max (start model): 125/125
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
inv.iter 2 ... chi² = 0.47 (dPhi = 38.88%) lam: 20.0


################################################################################
#                  Abort criterion reached: chi² <= 1 (0.47)                   #
################################################################################
```

**Comment by halbmy** [comment]
For synthetic experiments, you should always add noise to the data (done by adding the outcommented arguments in the `simulate()` call) and remove the `estimateError()` call (because `simulate` will do this when adding noise).

**Reply by TheNablaOperator** [reply]
Thanks for reply.
The result is a already with added noise, as suggested by @carsten-forty2  above. 
Currently I run the forward model: 
```
survey_noisy = ert.simulate(mesh, 
                   res=resmap, 
                   scheme=shm, 
                   verbose=True, 
                   noiseAbs=1e-6, seed=1337)
```

If I remove `survey_noisy["err"] = ert.estimateError(survey_noisy)` I get an error...
`survey_noisy["err"]` seems to be empty... Obviously not filled by `simulate()` ...?

Part of the error log:
```
16/03/22 - 12:18:50 - pyGIMLi - INFO - Found 1 regions.
16/03/22 - 12:18:50 - pyGIMLi - ERROR - <class 'pygimli.physics.ert.ertManager.ERTManager'>.checkError(C:\Users\benjaminjauk\.conda\envs\pg2\lib\site-packages\pygimli\frameworks\methodManager.py:340)
Datacontainer have no "err" values. Fallback set to 0.01
16/03/22 - 12:18:50 - pyGIMLi - CRITICAL - <class 'pygimli.physics.ert.ertManager.ERTManager'>._ensureError(C:\Users\benjaminjauk\.conda\envs\pg2\lib\site-packages\pygimli\frameworks\methodManager.py:359)
All error values need to be larger then 0. Either give and err argument or fill dataContainer  with a valid 'err'  0.0 0.0
```

**Reply by halbmy** [reply]
You need a (relative) noise level `noiseLevel` (e.g. 0.03) to trigger adding noise (and creating the `err` field).
A purely absolute noise (as voltage) does not make too much sense.

**Reply by TheNablaOperator** [reply]
Aah ok, ups! 
Yes works now.

**Comment by halbmy** [comment]
Right now you are fitting the data twice as good as the error model (3% by default). Sounds ok, but the data spread (100 to 184 Ohmm) is not very large. To decide whether the model smoothness is ok, you need noise.

**Reply by TheNablaOperator** [reply]
The tank (100 Ohm.m) contains a block (1000 Ohm.m).
I get the spread according to the forward simulation...

Side note:
If I simulate the same model not as tank, but as world:
The spread is: 76.27 to 189 Ohm.m
Compared to my old data (forward simulated with Geotomo Res3DMod) the spread was: 103 to 203 Ohm.m (but I don't know exactly how it works since it's not open Source...) For that reason I changed to pg...

**Reply by halbmy** [reply]
Sounds reasonable considering the anomaly is not huge and you'll have a mean value between 100 and 1000 according to the sensitivity.
